### PR TITLE
Fixing runtime errors encountered using python3

### PR DESCRIPTION
**Pull request by dedominic-veeder**
I couldn't get this program to work without significant changes. I made it at least successfully respond though I'm not sure it worked as my authcode was consumed and lost the output at the end.

### CODE REVIEW

**Comment by dedominic-veeder** [comment]
I ended up getting Tech support to generate a new auth code and this time I had to base64 encode the encrypted client nonce??? so I mgiht have to change this pr.

anyhow I have the key, token, expiration and mac but I don't know how to use it.

**Comment by dlenski** [comment]
> I couldn't get this program to work without significant changes. I made it at least successfully respond though I'm not sure it worked as my authcode was consumed and lost the output at the end.

Hmmm… what version of Py3 are you using? And what version of the [PyCryptoDome libraries](https://pycryptodome.readthedocs.io)? With Python 3.5.2 and PyCryptoDome 3.4.7 under Linux, it worked fine for me.

> anyhow I have the key, token, expiration and mac but I don't know how to use it.

You and me both :wink: . That's the problem here… even once we have the unencrypted key/token/expiration/mac, we still don't know how the official software converts all that into the seed for the token. https://github.com/cernekee/stoken/issues/28

File: client.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def parse_ServerResponse(self, response):
         assert x.tag == '{http://schemas.xmlsoap.org/soap/envelope/}Envelope'
         r = x.find('.//{http://ctkipservice.rsasecurity.com}' + outer)
         ad = r.find('{http://ctkipservice.rsasecurity.com}AuthData')
-        assert ad.text == self.auth == response.headers.get('Authorization')
+        assert ad.text == self.auth #== response.headers.get('Authorization')
```
Comment (cemeyer) [inline client.py:54]: @dlenski I also encountered an assertion failure here.  The response was of the form:

```
<?xml version="1.0" encoding="UTF-8"?>
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <soapenv:Body>
    <ServerResponse xmlns="http://ctkipservice.rsasecurity.com">
      <AuthData>XXX_ACTIVATION_CODE</AuthData>
      <ProvisioningData>XXX_Base64_of_another_xml_document_specifying_provisioning_version</ProvisioningData>
      <Response> XXX crapload of base64 here</Response>
    </ServerResponse>
  </soapenv:Body>
</soapenv:Envelope>
```

It looks like `ad.text` matches `self.auth` (i.e., supplied activation code).  Unfortunately I don't know what `response.headers.get('Authorization')` contained.

(2nd trial: the latter is just `None`.)

File: client.py [hunk 2]
```diff
@@ -84,30 +84,30 @@ def parse_ServerResponse(self, response):
 if args.verbose:
     print(res1)
 
-session_id = res1.getroot().attrib['SessionID']
```
Comment (cemeyer) [inline client.py:87]: I think this is also needed -- ET.fromstring, at least in py36, returns an Element, not an ElementTree

File: client.py [hunk 3]
```diff
@@ -84,30 +84,30 @@ def parse_ServerResponse(self, response):
...
+session_id = res1.attrib['SessionID']
 k = res1.find('.//{http://www.w3.org/2000/09/xmldsig#}RSAKeyValue')
 mod = number.bytes_to_long(d64sb(k.find(
   '{http://www.w3.org/2000/09/xmldsig#}Modulus').text))
 exp = number.bytes_to_long(d64sb(k.find(
   '{http://www.w3.org/2000/09/xmldsig#}Exponent').text))
-pubk = RSA.construct({'n': mod, 'e': exp})
+pubk = RSA.construct( (int(mod), int(exp)) )
```
Comment (cemeyer) [inline client.py:93]: Yeah, I needed this too.  The `int()`s are gratuitous but RSA.construct takes a tuple now.

File: client.py [hunk 4]
```diff
@@ -84,30 +84,30 @@ def parse_ServerResponse(self, response):
...
 pl = res1.find('.//Payload')
 server_nonce = d64sb(pl.find('Nonce').text)
 
 print("Got server nonce and RSA pubkey:\n{}\n{}".format(
     hexlify(server_nonce), pubk.exportKey('PEM').decode()))
 
 # generate and encrypt client nonce
-client_nonce = random.getrandbits(16*8)
+client_nonce = bytearray([random.getrandbits(8) for i in range(16)])
```
Comment (cemeyer) [inline client.py:101]: ditto -- cipher.encrypt() doesn't take a single int

File: client.py [hunk 5]
```diff
@@ -84,30 +84,30 @@ def parse_ServerResponse(self, response):
...
-encrypted_client_nonce = cipher.encrypt(client_nonce)
+encrypted_client_nonce = cipher.encrypt(hexlify(client_nonce))
 print("Generated client nonce:\n\tplaintext: {}\n\tencrypted: {}".format(
     hexlify(client_nonce), hexlify(encrypted_client_nonce)))
 
 # send second request
 req2_filled = req2_tmpl.format(
-  session_id=session_id, encrypted_client_nonce=encrypted_client_nonce,
-  server_nonce=server_nonce)
+  session_id=session_id, encrypted_client_nonce=hexlify(encrypted_client_nonce).decode(),
```
Comment (cemeyer) [inline client.py:109]: ditto

File: client.py [hunk 6]
```diff
@@ -84,30 +84,30 @@ def parse_ServerResponse(self, response):
...
 
 print("Got server nonce and RSA pubkey:\n{}\n{}".format(
     hexlify(server_nonce), pubk.exportKey('PEM').decode()))
 
 # generate and encrypt client nonce
-client_nonce = random.getrandbits(16*8)
+client_nonce = bytearray([random.getrandbits(8) for i in range(16)])
 cipher = PKCS1_OAEP.new(pubk)
-encrypted_client_nonce = cipher.encrypt(client_nonce)
+encrypted_client_nonce = cipher.encrypt(hexlify(client_nonce))
```
Comment (cemeyer) [inline client.py:103]: with or without this I get "Failed to process CT-KIP clientNonceRequest. Status code = MalformedRequest"

hmm

File: client.py (hunk 1 above)
Comment (dlenski) [reply client.py:54]: Thanks, that's interesting. "My" server duplicated the `authdata` tag in the `Authorization` header, but I guess others don't.

File: client.py (hunk 3 above)
Comment (dlenski) [reply client.py:93]: Thanks, do you know in what version of PyCryptoDome this changed?

https://pycryptodome.readthedocs.io/en/latest/src/public_key/rsa.html#Crypto.PublicKey.RSA.construct

File: client.py (hunk 2 above)
Comment (dlenski) [reply client.py:87]: Thanks. Appears to be the same back to at least [v2.7](https://docs.python.org/2.7/library/xml.etree.elementtree.html) and [v3.3](https://docs.python.org/3.3/library/xml.etree.elementtree.html)… not sure how how this worked for me :frowning_face:

(Should be replaced with `session_id = res1.attrib['SessionID']` in that case, not deleted.)

**Comment by dlenski** [comment]
@cemeyer and @dedominic-veeder, could you tell me exactly which versions of Python and PyCryptoDome you're running?

**Comment by cemeyer** [comment]
> @cemeyer and @dedominic-veeder, could you tell me exactly which versions of Python and PyCryptoDome you're running?

Python 3.6.7, PyCryptoDome 3.7.2

File: client.py (hunk 2 above)
Comment (cemeyer) [reply client.py:87]: Right, the following diff line is the replace.  I clicked the wrong line / github has poor code review UI :-P.

File: client.py (hunk 5 above)
Comment (cemeyer) [reply client.py:109]: Hm, I wonder if these should be base64, like the server nonce is sent?

Comment (cemeyer) [reply client.py:109]: Well that's fascinating, the error changed from Malformed Request to just "Abort"

Comment (dlenski) [reply client.py:109]: Yeah, it's definitely b64'ed :man_facepalming: 

I'm not sure how I borked this… in the `fakeserver.py` code it's clearly b64'ed, *not* hexlified:
https://github.com/dlenski/rsa_ct_kip/blob/master/fakeserver.py#L154

I did a *lot* more testing of `fakeserver.py` than of `client.py`, due to the fact that a "real" server is required to test from the client side.

If you get the client code to talk to the fake server, then I'm pretty confident that it'll talk to the real server as well. Sorry for wasting your time with incomplete code here…

Comment (cemeyer) [reply client.py:109]: no worries

did you find https://tools.ietf.org/html/draft-doherty-ct-kip-ws-00 ?  It has some examples at the end

Comment (cemeyer) [reply client.py:109]: http://dpaste.com/0WGQS25#wrap demangled somewhat.  i'll play with fakeclient <-> fakeserver :P

Comment (dlenski) [reply client.py:109]: Interesting. The exchanges I captured definitely _did not_ have the x509 certificate embedded.

Here's a complete transcript of the one I saved, de-SOAPified: http://dpaste.com/1P44DBX

Comment (cemeyer) [reply client.py:109]: Yes, mine look more like that too.  No x509 embedded.

Comment (dlenski) [reply client.py]: > Yes, mine look more like that too. No x509 embedded.

So where did you get the one you pasted…?

**Pull request merged**